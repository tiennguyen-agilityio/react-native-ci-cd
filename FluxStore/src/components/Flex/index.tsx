import {ReactNode, memo} from 'react';
import {StyleProp, StyleSheet, View, ViewStyle} from 'react-native';

import {FlexOptions} from '@/interfaces';

const flexDirection = StyleSheet.create({
  row: {
    flexDirection: 'row',
  },
  column: {
    flexDirection: 'column',
  },
  'row-reverse': {
    flexDirection: 'row-reverse',
  },
  'column-reverse': {
    flexDirection: 'column-reverse',
  },
});

const flexJustify = StyleSheet.create({
  start: {
    justifyContent: 'flex-start',
  },
  end: {
    justifyContent: 'flex-end',
  },
  center: {
    justifyContent: 'center',
  },
  between: {
    justifyContent: 'space-between',
  },
  around: {
    justifyContent: 'space-around',
  },
  evenly: {
    justifyContent: 'space-evenly',
  },
});

const flexAlign = StyleSheet.create({
  start: {
    alignItems: 'flex-start',
  },
  end: {
    alignItems: 'flex-end',
  },
  center: {
    alignItems: 'center',
  },
  stretch: {
    alignItems: 'stretch',
  },
  baseline: {
    alignItems: 'baseline',
  },
});

const flexWrap = StyleSheet.create({
  wrap: {
    flexWrap: 'wrap',
  },
  nowrap: {
    flexWrap: 'nowrap',
  },
  'wrap-reverse': {
    flexWrap: 'wrap-reverse',
  },
});

interface FlexProps extends FlexOptions {
  children?: ReactNode;
  style?: StyleProp<ViewStyle>;
}

const Flex = ({
  direction = 'column',
  justify = 'start',
  align = 'stretch',
  wrap = 'nowrap',
  children,
  style,
  ...props
}: FlexProps) => {
  return (
    <View
      testID="flex"
      style={[
        flexDirection[direction],
        flexJustify[justify],
        flexAlign[align],
        flexWrap[wrap],
        style,
        props,
      ]}>
      {children}
    </View>
  );
};

export default memo(Flex);
