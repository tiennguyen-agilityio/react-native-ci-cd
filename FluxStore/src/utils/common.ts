import {Cart} from '@/interfaces';

/**
 * Formats a number into a localized string with appropriate decimal places.
 *
 * - If the number is an integer, it returns the number with thousands separators (e.g., 1000 -> "1,000").
 * - If the number is a float, it formats it to 2 decimal places (e.g., 1234.5 -> "1,234.50").
 * - If the number is falsy (e.g., 0, null, undefined), it returns 0.
 *
 * @param number - The numeric value to format
 * @returns A formatted string or 0
 */
export const formatAmount = (number: number) => {
  if (number) {
    if (Number.isInteger(number)) {
      return number.toLocaleString('en-US');
    } else {
      return number.toLocaleString('en-US', {
        maximumFractionDigits: 2,
        minimumFractionDigits: 2,
      });
    }
  }
  return 0;
};

/**
 * Flattens paginated data into a single array.
 *
 * - Useful for infinite queries or paginated API responses.
 * - Accepts an array of pages (e.g., [[...], [...], ...]) and flattens them into a single array.
 *
 * @template T - The type of items in the pages
 * @param pages - An array of arrays (pages) containing items of type T
 * @returns A flat array of all items from all pages
 */
export const getData = <T>(pages = []) => {
  let result: T[] = [];

  pages?.map(item => {
    result = result.concat(item);
  });

  return result;
};

/**
 * Converts an object of filters into a URL query string.
 *
 * - Filters out keys with `undefined`, `null`, or empty string values.
 * - Trims string values and URL-encodes both keys and values.
 *
 * @param filters - An object containing key-value pairs to be serialized into a query string
 * @returns A query string (e.g., "key1=value1&key2=value2")
 */
export const getQueryString = (filters: object = {}): string =>
  Object.entries(filters)
    .filter(([, v]) => v !== undefined && v !== null && v !== '')
    .map(
      ([k, v]) =>
        `${encodeURIComponent(k)}=${encodeURIComponent(typeof v === 'string' ? v.trim() : v)}`,
    )
    .join('&');

/**
 * Calculates the total price of checked items in the cart.
 *
 * - Filters cart items where `isChecked` is true.
 * - Multiplies each item's price by its quantity and sums the total.
 * - Formats the total amount as a localized string.
 *
 * @param carts - Array of cart items
 * @returns Formatted total price as a string
 */
export const calcTotalPrice = (carts: Cart[]) => {
  const total = carts
    .filter(c => c.isChecked)
    .reduce((sum, c) => sum + c.product.price * c.quantity, 0);

  return formatAmount(total);
};
