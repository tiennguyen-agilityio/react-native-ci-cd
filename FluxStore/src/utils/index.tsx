export * from './render';
export * from './common';
export * from './time';
export * from './notification';
export * from './performance';
