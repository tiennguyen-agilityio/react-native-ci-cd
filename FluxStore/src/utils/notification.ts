import {getApp} from '@react-native-firebase/app';
import {getMessaging} from '@react-native-firebase/messaging';
import notifee, {AndroidImportance} from '@notifee/react-native';
import {CHANNEL_NOTIFICATION, LINKING_URLS} from '@/constants';
import {DeepLinkData} from '@/interfaces';

/**
 * Builds a deep link URL from the provided notification data.
 *
 * This function extracts the `navigationId` and `postId` from the notification
 * data and constructs a URL based on these values. If the `navigationId` is
 * "browse", it returns the browse link. If a valid `postId` is provided, it
 * returns the product detail link for that post. Otherwise, it defaults to
 * the home link.
 *
 * @param data - An object containing the notification data, which may include
 * `navigationId` and `postId` properties.
 *
 * @returns A string representing the deep link URL based on the notification
 * data, or null if a URL cannot be constructed.
 */
export const buildDeepLinkFromNotificationData = (data?: DeepLinkData): string | null => {
  const {id, isProductList} = data || {};
  if (typeof id === 'string') {
    return LINKING_URLS.PRODUCT_DETAIL(id);
  } else if (isProductList) {
    return LINKING_URLS.PRODUCTS;
  }

  return LINKING_URLS.HOME;
};

/**
 * Retrieves the current device's FCM token for push notifications.
 *
 * Uses Firebase Cloud Messaging to get a unique token for the device, which can be
 * used to send targeted notifications from a server.
 *
 * @returns A Promise that resolves to the FCM token as a string, or `undefined` if it fails.
 */
export const getDeviceToken = async () => {
  try {
    const messagingInstance = getMessaging(getApp());
    const token = await messagingInstance.getToken();

    console.log('Token', token);
    return token;
  } catch (error) {
    console.log('Error getting FCM Token', error);
  }
};

/**
 * Creates a notification channel for Android.
 *
 * Android requires notification channels for push/local notifications.
 * This function sets up a high-importance default channel if on Android.
 * No-op on iOS, as channels are Android-specific.
 */
export const createNotificationChannel = async () => {
  // if (Platform.OS !== 'android') return;
  await notifee.createChannel({
    id: CHANNEL_NOTIFICATION.DEFAULT.id,
    name: CHANNEL_NOTIFICATION.DEFAULT.name,
    importance: AndroidImportance.HIGH,
  });
};
