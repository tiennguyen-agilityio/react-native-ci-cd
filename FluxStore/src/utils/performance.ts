import {User} from '@/interfaces';
import perf, {FirebasePerformanceTypes} from '@react-native-firebase/perf';
import crashlytics from '@react-native-firebase/crashlytics';

interface CustomTraceReturn {
  trace: FirebasePerformanceTypes.Trace;
  traceStop: () => Promise<null>;
}

/**
 * Starts a custom Firebase Performance Monitoring trace.
 *
 * - Creates a new performance trace with the given name.
 * - Starts the trace and returns a reference along with a stop function.
 *
 * @param name - The name of the custom trace.
 * @returns An object containing the trace instance and a `traceStop` method to stop it.
 */
export const customTrace = async (name: string): Promise<CustomTraceReturn> => {
  const trace = await perf().newTrace(name);
  await trace.start();

  return {
    trace,
    traceStop: () => trace.stop(),
  };
};

/**
 * Simulates a crash in the app and logs user info to Firebase Crashlytics.
 *
 * - Logs a message indicating a mock crash.
 * - Sets user ID and attributes (email, username) for Crashlytics context.
 * - Triggers a deliberate crash (useful for testing Crashlytics setup).
 *
 * @param user - The user object containing `id`, `email`, and `name` fields.
 * @returns A promise that resolves when the crash is triggered.
 */
export const mockCrashlyticLogin = async (user: User) => {
  crashlytics().log('User mock crash app.');
  await Promise.all([
    crashlytics().setUserId(user.id),
    crashlytics().setAttributes({
      email: user.email,
      username: user.name,
    }),
  ]);
  return crashlytics().crash();
};
