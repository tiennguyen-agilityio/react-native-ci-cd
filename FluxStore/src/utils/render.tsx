import Text from '@/components/Text';

export const RenderChildren = () => (
  <>
    <Text>Item 1</Text>
    <Text>Item 2</Text>
  </>
);
