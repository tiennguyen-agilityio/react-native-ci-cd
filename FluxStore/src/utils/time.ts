export const getShortTimeAgo = (timeStamp: number): string => {
  const diffMs = Date.now() - timeStamp;

  const diffMins = Math.floor(diffMs / (1000 * 60));
  if (diffMins < 60) return `${diffMins}m ago`;

  const diffHours = Math.floor(diffMins / 60);
  if (diffHours < 24) return `${diffHours}h ago`;

  const diffDays = Math.floor(diffHours / 24);
  return `${diffDays}d ago`;
};
