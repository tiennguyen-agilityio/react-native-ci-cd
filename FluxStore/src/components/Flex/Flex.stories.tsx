import {Meta, StoryObj} from '@storybook/react';

import {RenderChildren} from '@/utils';

import Flex from '.';
import {FlexOptions} from '@/interfaces';

const meta: Meta<typeof Flex> = {
  title: 'Components/Flex',
  component: Flex,
};

export default meta;

type Story = StoryObj<typeof Flex>;

const defaultProps = {
  gap: 5,
};

const Render = (props: FlexOptions) => {
  return (
    <Flex {...defaultProps} {...props}>
      <RenderChildren />
    </Flex>
  );
};

export const Default: Story = {
  render: Render,
};

export const Row: Story = {
  render: () => Render({direction: 'row'}),
};
