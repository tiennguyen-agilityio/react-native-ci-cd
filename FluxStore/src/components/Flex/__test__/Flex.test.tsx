import {render, screen} from '@testing-library/react-native';

import Flex from '..';
import {RenderChildren} from '@/utils';

describe('Flex', () => {
  it('should render component', () => {
    const {toJSON} = render(
      <Flex>
        <RenderChildren />
      </Flex>,
    );

    expect(toJSON()).toMatchSnapshot();
  });

  it('should render component with direction row', () => {
    const {toJSON} = render(
      <Flex {...{direction: 'row'}}>
        <RenderChildren />
      </Flex>,
    );

    expect(toJSON()).toMatchSnapshot();
  });

  it('should render component with align items stretch', () => {
    const {toJSON} = render(
      <Flex {...{align: 'stretch'}}>
        <RenderChildren />
      </Flex>,
    );

    expect(toJSON()).toMatchSnapshot();
  });

  it('should render component with empty children', () => {
    const {toJSON} = render(<Flex />);
    expect(toJSON()).toMatchSnapshot();
  });

  test('Should render children elements', () => {
    render(
      <Flex {...{align: 'stretch'}}>
        <RenderChildren />
      </Flex>,
    );
    const {children} = screen.getByTestId('flex');

    expect(children).toHaveLength(1);
  });
});
