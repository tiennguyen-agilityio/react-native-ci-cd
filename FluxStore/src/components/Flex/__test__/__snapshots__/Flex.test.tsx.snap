// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flex should render component 1`] = `
<View
  style={
    [
      {
        "flexDirection": "column",
      },
      {
        "justifyContent": "flex-start",
      },
      {
        "alignItems": "stretch",
      },
      {
        "flexWrap": "nowrap",
      },
      undefined,
      {},
    ]
  }
  testID="flex"
>
  <Text
    style={
      [
        {
          "letterSpacing": 0,
          "textAlign": "left",
          "textAlignVertical": "center",
        },
        {
          "color": "#737680",
          "fontFamily": "ProductSansRegular",
          "fontSize": 14,
        },
        {},
        undefined,
      ]
    }
  >
    Item 1
  </Text>
  <Text
    style={
      [
        {
          "letterSpacing": 0,
          "textAlign": "left",
          "textAlignVertical": "center",
        },
        {
          "color": "#737680",
          "fontFamily": "ProductSansRegular",
          "fontSize": 14,
        },
        {},
        undefined,
      ]
    }
  >
    Item 2
  </Text>
</View>
`;

exports[`Flex should render component with align items stretch 1`] = `
<View
  style={
    [
      {
        "flexDirection": "column",
      },
      {
        "justifyContent": "flex-start",
      },
      {
        "alignItems": "stretch",
      },
      {
        "flexWrap": "nowrap",
      },
      undefined,
      {},
    ]
  }
  testID="flex"
>
  <Text
    style={
      [
        {
          "letterSpacing": 0,
          "textAlign": "left",
          "textAlignVertical": "center",
        },
        {
          "color": "#737680",
          "fontFamily": "ProductSansRegular",
          "fontSize": 14,
        },
        {},
        undefined,
      ]
    }
  >
    Item 1
  </Text>
  <Text
    style={
      [
        {
          "letterSpacing": 0,
          "textAlign": "left",
          "textAlignVertical": "center",
        },
        {
          "color": "#737680",
          "fontFamily": "ProductSansRegular",
          "fontSize": 14,
        },
        {},
        undefined,
      ]
    }
  >
    Item 2
  </Text>
</View>
`;

exports[`Flex should render component with direction row 1`] = `
<View
  style={
    [
      {
        "flexDirection": "row",
      },
      {
        "justifyContent": "flex-start",
      },
      {
        "alignItems": "stretch",
      },
      {
        "flexWrap": "nowrap",
      },
      undefined,
      {},
    ]
  }
  testID="flex"
>
  <Text
    style={
      [
        {
          "letterSpacing": 0,
          "textAlign": "left",
          "textAlignVertical": "center",
        },
        {
          "color": "#737680",
          "fontFamily": "ProductSansRegular",
          "fontSize": 14,
        },
        {},
        undefined,
      ]
    }
  >
    Item 1
  </Text>
  <Text
    style={
      [
        {
          "letterSpacing": 0,
          "textAlign": "left",
          "textAlignVertical": "center",
        },
        {
          "color": "#737680",
          "fontFamily": "ProductSansRegular",
          "fontSize": 14,
        },
        {},
        undefined,
      ]
    }
  >
    Item 2
  </Text>
</View>
`;

exports[`Flex should render component with empty children 1`] = `
<View
  style={
    [
      {
        "flexDirection": "column",
      },
      {
        "justifyContent": "flex-start",
      },
      {
        "alignItems": "stretch",
      },
      {
        "flexWrap": "nowrap",
      },
      undefined,
      {},
    ]
  }
  testID="flex"
/>
`;
